Crash when using capability member in function params

## Compiler version

3.8.3

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```Scala
import scala.language.experimental.captureChecking

object Obj {
  type C^ = {caps.any}
  def foo(a: Any^{C}): Unit = ()
}
```

## Output (click arrow to expand)
<!--
Note: to get a full crash log when using scala-cli, run compilation with the flag `--server=false`
-->
<details>

```scala
  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/scala/scala3/issues/new/choose
  For non-enriched exceptions, compile with -Xno-enrich-error-messages.


     while compiling: /home/pc/lab/fyp/exp/src/main/scala/Main.scala
        during phase: cc
                mode: Mode(ImplicitsEnabled)
     library version: version 3.8.3
    compiler version: version 3.8.3
            settings: -classpath /home/pc/lab/fyp/exp/target/scala-3.8.3/classes:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.8.3/scala3-library_3-3.8.3.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/3.8.3/scala-library-3.8.3.jar -d /home/pc/lab/fyp/exp/target/scala-3.8.3/classes

[error] ## Exception when compiling 2 sources to /home/pc/lab/fyp/exp/target/scala-3.8.3/classes
[error] java.lang.AssertionError: assertion failed: bimap GlobalToLocalCap should map Obj.C to a capability, but result = {any}
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:10)
[error] dotty.tools.dotc.core.Types$BiTypeMap.mapCapability(Types.scala:6253)
[error] dotty.tools.dotc.core.Types$BiTypeMap.mapCapability$(Types.scala:6237)
[error] dotty.tools.dotc.cc.Capabilities$GlobalCapToLocal.mapCapability(Capability.scala:1088)
[error] dotty.tools.dotc.cc.CaptureSet.$anonfun$2(CaptureSet.scala:411)
[error] dotty.tools.dotc.util.SimpleIdentitySet$Set1.map(SimpleIdentitySet.scala:98)
[error] dotty.tools.dotc.cc.CaptureSet.map(CaptureSet.scala:411)
[error] dotty.tools.dotc.core.Types$TypeMap.mapCapturingType(Types.scala:6354)
[error] dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:6466)
[error] dotty.tools.dotc.cc.Capabilities$GlobalCapToLocal.apply(Capability.scala:1073)
[error] dotty.tools.dotc.cc.Capabilities$.op$proxy3$1(Capability.scala:1120)
[error] dotty.tools.dotc.cc.Capabilities$.globalCapToLocal(Capability.scala:1120)
[error] dotty.tools.dotc.cc.Setup.dotty$tools$dotc$cc$Setup$$transformExplicitType(Setup.scala:521)
[error] dotty.tools.dotc.cc.Setup$$anon$2.transformTT(Setup.scala:552)
[error] dotty.tools.dotc.cc.Setup$$anon$2.transformResultType(Setup.scala:563)
[error] dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:609)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1359)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
[error] dotty.tools.dotc.cc.Setup$$anon$2.traverse$$anonfun$3(Setup.scala:601)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:327)
[error] dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:601)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1359)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1360)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1360)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1360)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1778)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1820)
[error] dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:648)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1819)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1819)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1775)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1820)
[error] dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:624)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1359)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1360)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1357)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
[error] dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1784)
[error] dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1820)
[error] dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:648)
[error] dotty.tools.dotc.cc.Setup.setupUnit(Setup.scala:968)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.checkUnit(CheckCaptures.scala:2222)
[error] dotty.tools.dotc.transform.Recheck.run(Recheck.scala:154)
[error] dotty.tools.dotc.cc.CheckCaptures.run(CheckCaptures.scala:276)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:406)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:327)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:399)
[error] dotty.tools.dotc.transform.Recheck.runOn(Recheck.scala:158)
[error] dotty.tools.dotc.cc.CheckCaptures.runOn(CheckCaptures.scala:272)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:380)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1324)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:373)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$2(Run.scala:420)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$adapted$1(Run.scala:420)
[error] scala.Function0.apply$mcV$sp(Function0.scala:42)
[error] dotty.tools.dotc.Run.showProgress(Run.scala:482)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:420)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:432)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:432)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:319)
[error] dotty.tools.dotc.Run.compile(Run.scala:304)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:196)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:252)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:186)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:166)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:241)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:166)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:214)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:542)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:542)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$3(Incremental.scala:182)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$3$adapted(Incremental.scala:180)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:458)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:264)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:413)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:500)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:400)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:208)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:542)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:496)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:433)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2343)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2293)
[error] sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:41)
[error] sbt.internal.io.Retry$.sbt$internal$io$Retry$$impl(Retry.scala:114)
[error] sbt.internal.io.Retry$.io(Retry.scala:102)
[error] sbt.internal.io.Retry$.io(Retry.scala:78)
[error] sbt.internal.io.Retry$.io(Retry.scala:67)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:41)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2291)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:63)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:69)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:283)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
[error] sbt.Execute.work(Execute.scala:292)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:283)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:65)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[error] java.base/java.lang.Thread.run(Thread.java:1583)
[error]            
[error] stack trace is suppressed; run last Compile / compileIncremental for the full output
[error] (Compile / compileIncremental) java.lang.AssertionError: assertion failed: bimap GlobalToLocalCap should map Obj.C to a capability, but result = {any}
```
</details>

--- a/compiler/src/dotty/tools/dotc/cc/Capability.scala
+++ b/compiler/src/dotty/tools/dotc/cc/Capability.scala
@@ -1179,7 +1179,11 @@ object Capabilities:
   // ---------- Maps between different kinds of root capabilities -----------------
 
   /** Map each occurrence of `caps.any` to a different LocalCap instance
-   *  Exception: CapSet^ stays as it is.
+   *  Exception: Capture set references like an `X` with `type X^ = {caps.any}`
+   *  are kept as they are. The `caps.any` in the alias is freshened once when
+   *  the alias definition itself is set up, so all uses of `X` refer to that
+   *  same instance. Following the alias in `mapCapability` instead would map
+   *  `X` to a capture set, which cannot be an element of a mapped set (i26180).
    */
   class GlobalCapToLocal(origin: Origin)(using Context) extends BiTypeMap, FollowAliasesMap:
     thisMap =>
@@ -1212,6 +1216,7 @@ object Capabilities:
         // `cap` to reach capabilities, so invariant occurrences were never
         // connected to their environment.
         LocalCap(origin, atInvariantPos = variance == 0)
+      case c: TypeRef if c.derivesFromCapSet => c
       case _ => super.mapCapability(c)
 
     override def fuse(next: BiTypeMap)(using Context) = next match
@@ -1227,6 +1232,7 @@ object Capabilities:
 
       override def mapCapability(c: Capability): Capability = c match
         case c: LocalCap if globalizes(c) => GlobalAny
+        case c: TypeRef if c.derivesFromCapSet => c
         case _ => super.mapCapability(c)
 
       def inverse = thisMap

--- a/tests/neg-custom-args/captures/i26180.scala
+++ b/tests/neg-custom-args/captures/i26180.scala
@@ -1,0 +1,15 @@
+import scala.language.experimental.captureChecking
+
+class IO
+
+object Obj:
+  type C^ = {caps.any}
+  def bar(a: IO^{C}): IO^{C} = a
+
+object Abs:
+  type C^ <: {caps.any}
+  def bar(a: IO^{C}): IO^{C} = a
+
+def test(io: IO^): Unit =
+  val x = Obj.bar(io) // error: local capability io cannot flow into the fixed set Obj.C
+  val y = Abs.bar(io) // error: same for an abstract capture-set member

--- a/tests/pos-custom-args/captures/i26180.scala
+++ b/tests/pos-custom-args/captures/i26180.scala
@@ -1,0 +1,24 @@
+import scala.language.experimental.captureChecking
+
+class IO
+
+object Obj:
+  type C^ = {caps.any}
+  def foo(a: Any^{C}): Unit = ()            // was: crash in GlobalCapToLocal
+  def bar(): Any^{C} = ???                  // result position
+  val v: Any^{C} = ???                      // val type
+  def baz(f: (Any^{C}) -> Unit): Unit = ()  // was: crash in GlobalCapToLocal.Inverse
+  def qux(xs: List[Any^{C}]): Unit = ()     // boxed position
+
+object Mixed:
+  val x: IO^ = ???
+  type C^ = {caps.any, x}
+  def foo(a: Any^{C}): Unit = ()
+
+class Cls:
+  type C^ = {caps.any}
+  def foo(a: Any^{C}): Unit = ()
+  def bar(a: Any^{this.C}): Unit = ()
+
+def test(io: IO^): Unit =
+  Obj.foo(io)

